Druid 31.0.0 release notes (#17092)
apache/druid: docs/release-info/release-notes.md

Reviewer: We need a section on compaction in the data management section.

Author: Added - what to put in it?

Reviewer 2: In compaction section add more details about msq autocompact etc

This section contains important information about new and existing features.

### Compaction features

Druid now supports the following features:

> - Compaction scheduler with greater flexibility and control over when and what to compact.

Reviewer 3: ```suggestion
- Auto-compaction supervisor with greater flexibility and control over when and what to compact.
```

- MSQ-based compaction for performant compaction jobs.

See [Automatic compaction](../data-management/automatic-compaction.md) for details.

> Concurrent compaction is now generally available.

Reviewer 3: ```suggestion
Concurrent compaction is now generally available (GA).
```

- Improved Overlord performance by reducing redundant calls in SQL statements [#16839](https://github.com/apache/druid/pull/16839)
- Improved `CustomExceptionMapper` so that it returns a correct failure message [#17016](https://github.com/apache/druid/pull/17016)
- Fixed NPE in `CompactSegments` [#16713](https://github.com/apache/druid/pull/16713)
- Fixed Parquet reader to ensure that Druid reads the required columns for a filter from the Parquet data files [#16874](https://github.com/apache/druid/pull/16874)

### SQL-based ingestion

#### Optimized S3 storage writing for MSQ durable storage

For queries that use the MSQ task engine and write their output to S3 as durable storage, uploading chunks of data is now faster.

[#16481](https://github.com/apache/druid/pull/16481)

#### Improved lookup performance

Improved lookup performance for queries that use the MSQ task engine by only loading required lookups. This applies to both ingestion and querying.

[#16358](https://github.com/apache/druid/pull/16358)

#### Other SQL-based ingestion improvements

> - Added query context parameter `maxNumSegments` [#16637](https://github.com/apache/druid/pull/16637)

Reviewer 4: No need to call this out as it was intentionally undocumented

- Improved query filtering to correctly process cases where both an IN expression and an equality (`=`) filter are applied to the same string value [#16597](https://github.com/apache/druid/pull/16597)
- Improved the speed of SQL IN queries that use the SCALAR_IN_ARRAY function [#16388](https://github.com/apache/druid/pull/16388)
- Updated the deserialization of dimensions in GROUP BY queries to operate on all dimensions at once rather than deserializing individual dimensions [#16740](https://github.com/apache/druid/pull/16740)
- Fixed an issue that caused `maxSubqueryBytes` to fail when segments had missing columns [#16619](https://github.com/apache/druid/pull/16619)
- Fixed an issue with the array type selector that caused the array aggregation over window frame to fail [#16653](https://github.com/apache/druid/pull/16653)
- Fixed support for native window queries without a group by clause [#16753](https://github.com/apache/druid/pull/16753)

### Cluster management

#### Added cluster configuration for multivalue handling

> Added  an optional cluster configuration property `druid.indexing.formats.stringMultiValueHandlingMode` which overrides `SORTED_SET` for string dimensions. Possible settings are `SORTED_SET` (default), `SORTED_ARRAY`, `ARRAY`.

Reviewer 4: nit: extra space & wording:
```suggestion
Added an optional cluster configuration property `druid.indexing.formats.stringMultiValueHandlingMode` which overrides the default mode `SORTED_SET` used for string dimensions. Possible values are `SORTED_SET` (default), `SORTED_ARRAY`, `ARRAY`.
```

Reviewer 4: Also, I think it would be better to add a note here recommending that users migrate to array types instead of MVDs, and include a reference to the relevant arrays upgrade notes section.

[#16822](https://github.com/apache/druid/pull/16822)

#### API for cluster-level compaction configuration

Druid now offers an API to update the cluster-level compaction dynamic configuration:

```
POST /druid/coordinator/v1/config/compaction/cluster
```

Example payload:

```json
{
     "compactionTaskSlotRatio": 0.5,
     "maxCompactionTaskSlots": 10,
     "engine": "msq"
}
```

> This API deprecates the older API `/druid/coordinator/v1/config/compaction/taskslots`.

Reviewer 4: Add this to a separate "Deprecated" section to make it easier to identify all the deprecated items?

Reviewer 3: I think it is already included in the list of `Deprecations` at the top.

[#16803](https://github.com/apache/druid/pull/16803)

#### Guardrail for task payload size

The `druid.indexer.queue.maxTaskPayloadSize` config defines the maximum size in bytes of a task payload accepted by the Overlord service. The default is 60 MB.

[#16512](https://github.com/apache/druid/pull/16512)

#### Other cluster management improvements

- The startup script for Docker-based deployments now specifies the node type [#16282](https://github.com/apache/druid/pull/16282)
- The `druid.sh` script now uses the canonical hostname for `druid.host` by default. You can now set the following environment variable to use IP instead: `DRUID_SET_HOST_IP=1` [#16386](https://github.com/apache/druid/pull/16386)

### Data management

#### Compaction

[TBC]

> 

Reviewer 4: Can group these kill task improvements under a separate sub-section "Kill" similar to the above "Compaction" section.

> #### Different task locks for kill tasks (experimental)

Reviewer 4: Move this experimental feature to the end of the section?

Kill tasks can now use different types of locks, such as APPEND or REPLACE. This change is experimental and not recommended for produciton use.

> As part of this change, kill tasks no longer use the `markAsUnused` parameter.

Reviewer 4: This kill task parameter was [not](https://druid.apache.org/docs/latest/data-management/delete) documented, so perhaps we can just omit this.

[#16831](https://github.com/apache/druid/pull/16831)

#### Added indexer metrics

Added `indexer/task/failed/count` and `indexer/task/failed/count` metrics.

[#16829](https://github.com/apache/druid/pull/16829)

#### Added subquery metrics

Added `subquery/rows` and `subquery/bytes` metrics, which indicate the size of the results materialized on the heap.

[#16835](https://github.com/apache/druid/pull/16835)

#### Added segment loading rate metrics

Made the following changes to segment loading rate metrics:

- Added `LoadingRateTracker` which computes a moving average load rate based on
the last few GBs of successful segment loads.
- Added `expectedLoadTimeMillis` to `/druid/coordinator/v1/loadQueue?simple` API response.
- Updated the Coordinator to emit `segment/loading/rateKbps`.

[#16691](https://github.com/apache/druid/pull/16691)

### Extensions

#### Kubernetes improvements

When the fabric client is unable to communicate with the Kubernetes server, the shutdown logic now proceeds without populating the task status location, ensuring all data structures are cleaned up.

[#16711](https://github.com/apache/druid/pull/16711)

> #### `pjac4j` OIDC context

Reviewer 3: ```suggestion
#### `pac4j` OIDC context
```

The `pac4j` extension now returns the OIDC context for `AuthenticationResult`.

[#16109](https://github.com/apache/druid/pull/16109)

#### Middle Manager-less Druid improvements

The Dynamic Pod Template Selection feature enhances the K8s extension by enabling more flexible and dynamic selection of pod templates based on task properties, which can optimize resource utilization and task execution efficiency.

[#16510](https://github.com/apache/druid/pull/16510)

In MiddleManager-less ingestion, Druid adds the pod template name as an annotation to the created job or pod.

[#16772](https://github.com/apache/druid/pull/16772)

#### Configure case sensitivity for Iceberg

> You can now use the `caseSensitive` Boolean config to configure how Druid reads column names from Iceberg.

Reviewer 4: Perhaps mention that this is optional, and that the the Iceberg table scans has case sensitivity enabled by default.

[#16496](https://github.com/apache/druid/pull/16496)

#### Other extensions improvements

> - Added an informational entry to the supervisor log when task count is greater than the number of partitions [#16948](https://github.com/apache/druid/pull/16948)

Reviewer 3: ```suggestion
- Added an informational entry to the streaming supervisor log when task count is greater than the number of partitions [#16948](https://github.com/apache/druid/pull/16948)
```

This section contains important information about new and existing features.

> ### Compaction on MSQ

Author: updated

The MSQ task engine now supports compaction with the following exceptions:

- Updating the cluster-level default compaction engine isn't supported. The current default is `native`, which can be changed on a per-datasource basis. 
- Grouping on multi-value columns isn't supported. The flag `groupByEnableMultiValueUnnesting` is disabled. Only array-type columns are supported.
- `partitionsSpec` of type `HashedParititionsSpec` isn't supported. Only `DimensionRangePartitionsSpec` and `DynamicPartitionsSpec` are supported.
- `maxTotalRows` in `DynamicPartitionsSpec` isn't supported. Only `maxRowsPerSegment` works.
- Setting `rollup` to false in `granularitySpec` when `metricsSpec` is specified isn't supported. Only `rollup` set to true works with a non-empty `metricsSpec`.
- Aggregators in `metricsSpec` where `fieldName != name` aren't supported.

[#16291](https://github.com/apache/druid/pull/16291)

### Window functions are GA

> [Window functions](../querying/sql-window-functions.md) are now generally available.

Author: updated

![](31.0_copy_results_as_sql.png)

When you copy the results of the pictured query, you get the following query:

```sql
SELECT
  CAST("c1" AS VARCHAR) AS "channel",
  CAST("c2" AS VARCHAR) AS "cityName",
  DECODE_BASE64_COMPLEX('thetaSketch', "c3") AS "user_theta"
FROM (
  VALUES
  ('ca', NULL, 'AQMDAAA6zJOcUskA1pEMGA=='),
  ('de', NULL, 'AQMDAAA6zJP43ITYexvoEw=='),
  ('de', NULL, 'AQMDAAA6zJNtue8WOvrJdA=='),
  ('en', NULL, 'AQMDAAA6zJMruSUUqmzufg=='),
  ('en', NULL, 'AQMDAAA6zJM6dC5sW2sTEg=='),
  ('en', NULL, 'AQMDAAA6zJM6dC5sW2sTEg=='),
  ('en', NULL, 'AQMDAAA6zJPqjEoIBIGtDw=='),
  ('en', 'Caulfield', 'AQMDAAA6zJOOtGipKE6KIA=='),
  ('fa', NULL, 'AQMDAAA6zJM8nxZkoGPlLw=='),
  ('vi', NULL, 'AQMDAAA6zJMk4ZadSFqHJw==')
) AS "t" ("c1", "c2", "c3")
```

[#16458](https://github.com/apache/druid/pull/16458)

#### Search for a datasource

You can know search for datasources in the **Datasource** view. Previously, you had to find them manually.

[#16371](https://github.com/apache/druid/pull/16371)

#### Display formatted JSON

You can now display both raw and formatted JSON in tables, making the data easier to read and troubleshoot.

> [#16632](https://github.com/apache/druid/pull/16632)

Author: updated

- Added the ability to configure `serverQueryContext` to set the query context [#16868](https://github.com/apache/druid/pull/16868)
- Added column mapping information to the explain plan [#16598](https://github.com/apache/druid/pull/16598)
- Added the ability to initiate handoff for a supervisor [#16586](https://github.com/apache/druid/pull/16586)
- Added an option to `Use concurrent locks` and moved all insert and replace options to a separate submenu [#16899](https://github.com/apache/druid/pull/16899)
- Improved how the web console detects durable storage [#16493](https://github.com/apache/druid/pull/16493)
- Made the following web console improvements:
  - Added titles to action menus
  - Improved the query-based ingestion counter calculation
  - Removed the filter clause on `__time`
  - Fixed scrolling in the `loadRules` editor [#16735](https://github.com/apache/druid/pull/16735)
- Restored the default WHERE filter to auto-generated SQL queries [#16608](https://github.com/apache/druid/pull/16608)
- Fixed NPE due to null values in numeric columns [#16760](https://github.com/apache/druid/pull/16760)

### Ingestion

#### General ingestion

> #### Changed default for indexer task lock batch allocation wait time

Author: updated

The default value for `druid.indexer.tasklock.batchAllocationWaitTime` is now 0.

A segment allocation request is processed immediately unless there are already some requests queued before it. While in queue, a segment allocation request may get clubbed together with other similar requests into a batch to reduce load on the metadata store.

[#16578](https://github.com/apache/druid/pull/16578)

> #### API to fetch conflicting task locks

Author: Moved into the cluster management section, there is no API section.

You can use the new API, `/activeLocks` to view task locks for the following attributes:
* Datasource
* Priority
* Interval list

Use this API for debugging or to view lock contention or observe any concurrent locks.

#### Other general ingestion improvements

- Hadoop-based ingestion now works on Kubernetes deployments [#16726](https://github.com/apache/druid/pull/16726)
- Hadoop-based ingestion now has a Boolean config `useMaxMemoryEstimates` parameter, which controls how memory footprint gets estimated. The default is false, so that the behavior matches native JSON-based batch ingestion [#16280](https://github.com/apache/druid/pull/16280)
- Added `druid-parquet-extensions` to all example quickstart configurations [#16664](https://github.com/apache/druid/pull/16664)
- Added support for ingesting CSV format data into Kafka records when Kafka ingestion is enabled with `ioConfig.type = kafka` [#16630](https://github.com/apache/druid/pull/16630)
- Added logging for sketches on workers [#16697](https://github.com/apache/druid/pull/16697)

> - Added the ability to sort segments by dimensions other than `__time` [#16849](https://github.com/apache/druid/pull/16849) [#16958](https://github.com/apache/druid/pull/16958)

Author: There isn't a section on Storage improvements - added Storage as a section in Functional areas.

- Removed obsolete tasks `index_realtime` and `index_realtime_appenderator` tasks&mdash;you can no longer use these tasks to ingest data [#16602](https://github.com/apache/druid/pull/16602)
- Renamed `TaskStorageQueryAdapter` to `TaskQueryTool` and removed the `isAudited` method [#16750](https://github.com/apache/druid/pull/16750)
- Improved Overlord performance by reducing redundant calls in SQL statements [#16839](https://github.com/apache/druid/pull/16839)
- Improved `CustomExceptionMapper` so that it returns a correct failure message [#17016](https://github.com/apache/druid/pull/17016)
- Fixed NPE in `CompactSegments` [#16713](https://github.com/apache/druid/pull/16713)
- Fixed Parquet reader to ensure that Druid reads the required columns for a filter from the Parquet data files [#16874](https://github.com/apache/druid/pull/16874)

#### SQL-based ingestion

#### Added query context parameter `maxNumSegments`

Added a query context parameter `maxNumSegments` that acts as a guard rail on the maximum number of segments that can be generated by an ingestion task per time chunk. The default value is null (no restriction on the number of segments per time chunk).

> [#16637](https://github.com/apache/druid/pull/16637)

Author: updated

#### Improved lookup performance

Improved lookup performance for queries that use the MSQ task engine by only loading required lookups. This applies to both ingestion and querying.

[#16358](https://github.com/apache/druid/pull/16358)

#### Compression for complex columns

Compression is now available for all complex metric columns which don't have specialized implementations. You can configure the `complexMetricCompression` IndexSpec option to any compression strategy (lz4, zstd, etc). The default is uncompressed. This works for most complex columns except `compressed-big-decimal` and columns stored by first/last aggregators.

Note that enabling compression is not backwards compatible with Druid versions older than 31. Only enable it after comprehensive testing, to ensure no need for rollback.

> [#16863](https://github.com/apache/druid/pull/16863)

Author: updated

> #### Fixed partition boundary issues for arrays

Author: updated

Updated logic to fix incorrect query results for comparisons involving arrays.

[#16780](https://github.com/apache/druid/pull/16780)

##### Other SQL-based ingestion improvements

- Reduced memory usage when transferring sketches between the MSQ task engine controller and worker [#16269](https://github.com/apache/druid/pull/16269)
- Improved error handling when retrieving Avro schemas from registry [#16684](https://github.com/apache/druid/pull/16684)
- Fixed issues related to partitioning boundaries in the MSQ task engine's window functions [#16729](https://github.com/apache/druid/pull/16729)
- You can now pass a custom `DimensionSchema` map to MSQ query destination of type `DataSourceMSQDestination` instead of using the default values [#16864](https://github.com/apache/druid/pull/16864)

#### Streaming ingestion

> #### New API for exiting streaming task groups early

Author: updated

This new API does a best effort attempt to triggerthe handoff for specified task groups of a supervisor early:

`POST` `/druid/indexer/v1/supervisor/{supervisorId}/taskGroups/handoff`

[#16310](https://github.com/apache/druid/pull/16310)

##### Other streaming ingestion improvements

- Added a check for handing off upgraded real-time segments. This prevents data from being temporarily unavailable for queries during segment handoff [#16162](https://github.com/apache/druid/pull/16162)

> - Added a Kinesis input format for timestamp and payload parsing [#16813](https://github.com/apache/druid/pull/16813)

Author: updated

- Improved the user experience for autoscaling Kinesis. Switching to autoscaling based on max lag per shard from total lag for shard is now controlled by the `lagAggregate` config, defaulting to sum [#16334](https://github.com/apache/druid/pull/16334)
- Improved the Supervisor so that it doesn't change to a running state from idle if the Overlord restarts [#16844](https://github.com/apache/druid/pull/16844)

### Querying

> #### Optimized S3 storage writing for MSQ durable storage

Author: updated

For queries that use the MSQ task engine and write their output to S3 as durable storage, uploading chunks of data is now faster.

[#16481](https://github.com/apache/druid/pull/16481)

#### MSQ task engine improved

Improved the performance of the MSQ task engine so that it supports in-memory shuffles and cached data across queries.

> [#16168](https://github.com/apache/druid/pull/16168)

Author: removed

#### Enabled querying cold datasources

<!-- Need more information on CentralizedDatasourceSchema. Where and how do you enable this feature. -->

You can now query entirely cold datasources after you enable the `CentralizedDatasourceSchema` feature.

[#16676](https://github.com/apache/druid/pull/16676)

#### SQL DIV function

You can now use the SQL DIV function.

[#16464](https://github.com/apache/druid/pull/16464)

> #### Improved query filtering

Author: updated

Improved query filtering to correctly process cases where both an IN expression and an equality (`=`) filter are applied to the same string value. Previously, such queries might have returned an empty result set.

[#16597](https://github.com/apache/druid/pull/16597)

Improved the speed of SQL IN queries that use the SCALAR_IN_ARRAY function.

[#16388](https://github.com/apache/druid/pull/16388)

#### Modified equality and typed in filter behavior

Modified the behavior of using `EqualityFilter` and `TypedInFilter` to match numeric values (particularly DOUBLE) against string columns by casting strings for numerical comparison. This ensures more consistent Druid behavior when the `sqlUseBoundAndSelectors` context flag is set.

[#16593](https://github.com/apache/druid/pull/16593)

> #### Faster dimension deserialization on Brokers

Author: updated

Updated the deserialization of dimensions in GROUP BY queries to operate on all dimensions at once rather than deserializing individual dimensions.

[#16740](https://github.com/apache/druid/pull/16740)

> #### Updated representation for window frames

Author: Zoltan says this isn't a user-facing change. Removed.

The following examples compare the old and new representations.

**Example 1**

```
# old
frame: { peerType: "ROWS", lowUnbounded: true, lowOffset: 0, uppUnbounded: true, uppOffset: 0 }

# new
frame: { type: rows }
```

**Example 2**

```
# old
frame: { peerType: "ROWS", lowUnbounded: false, lowOffset: 0, uppUnbounded: false, uppOffset: 2 }

# new
frame: { type: rows, lowerOffset: 0, upperOffset: 2 }
```

**Example 3**

```
# old
frame: { peerType: "RANGE", lowUnbounded: true, lowOffset: 0, uppUnbounded: false, uppOffset: 0, orderBy: [{ column: l1, direction: ASC }] }

# new
frame: { type: group, upperOffset: 0, orderBy: [{ column: l1, direction: ASC }] }
```

[#16741](https://github.com/apache/druid/pull/16741)

#### Query guardrails

**Asked Sree for a better release note** Druid rejects queries with HAVING clauses that contain windowed expressions as a result of a Calcite dependency.

[#16742](https://github.com/apache/druid/pull/16742)

Druid blocks window queries that involve aggregation functions inside the window clause, when the window is included in SELECT.
The error message provides details on updating your query syntax.

> The validation was added as a result of a Calcite dependency.

Author: removed

[#16801](https://github.com/apache/druid/pull/16801)

A query that references aggregators called with unsupported distinct values now fails.

[#16770](https://github.com/apache/druid/pull/16770)

Druid now validates that a complex type aligns with the supported types when used with an aggregator.
Queries like `SELECT COUNT(DISTINCT columnName) FROM tableName WHERE columnValue = 'non-existing-value'` will also stop working with these changes.

[#16682](https://github.com/apache/druid/pull/16682)

> Druid prevents you from using DISTINCT or unsupported aggregations with window functions.

Author: updated

This section contains important information about new and existing features.

### Compaction features

Druid now supports the following features:

- Compaction scheduler with greater flexibility and control over when and what to compact.
- MSQ task engine-based compaction for more performant compaction jobs.

See [Automatic compaction](https://druid.apache.org/docs/latest/data-management/automatic-compaction/) for details.

Compaction tasks that take advantage of concurrent append and replace is now generally available.

[#16291](https://github.com/apache/druid/pull/16291)

### Window functions are GA

> [Window functions](https://druid.apache.org/docs/latest//querying/sql-window-functions.md) are now generally available in Druid's native engine and in the MSQ task engine.

Reviewer 5: The URL seems incorrect. `latest//querying` -> `latest/querying`

- Fixed a distinct sketches issue where Druid called `retainedKeys.firstKey()` twice when adding another sketch [#17184](https://github.com/apache/druid/pull/17184)
- Fixed a `WindowOperatorQueryFrameProcessor` issue where larger queries could reach the frame writer's capacity preventing it from outputting all of the result rows [#17209](https://github.com/apache/druid/pull/17209)
- Fixed native ingestion task failures during rolling upgrades from a version before Druid 30 [#17219](https://github.com/apache/druid/pull/17219)

### SQL-based ingestion

#### Optimized S3 storage writing for MSQ durable storage

For queries that use the MSQ task engine and write their output to S3 as durable storage, uploading chunks of data is now faster.

[#16481](https://github.com/apache/druid/pull/16481)

#### Improved lookup performance

Improved lookup performance for queries that use the MSQ task engine by only loading required lookups. This applies to both ingestion and querying.

[#16358](https://github.com/apache/druid/pull/16358)

> #### Other SQL-based ingestion improvements

Reviewer 5: Is this file expected to contain all PRs marked with milestone 31.0.0? For example, I don't see https://github.com/apache/druid/pull/16804 mentioned, is that expected?

Reviewer 2: Hey Akshat, we typically don't include bug fixes unless there's a specific reason to. It's just new features/improvements. There are currently some fixes in there that I'll remove as part of the final cleanup.

Reviewer 5: I see, thanks for the info!

> It looks like 16804 and 17141 didn't have the bug labeled applied. Was that intentional?

Nope. I don't have the access to update PR labels, but yes both those PRs are bug-fixes.

- Added the ability to use `useConcurrentLocks` in task context to determine task lock type [#17193](https://github.com/apache/druid/pull/17193)
- Reduced memory usage when transferring sketches between the MSQ task engine controller and worker [#16269](https://github.com/apache/druid/pull/16269)
- Improved error handling when retrieving Avro schemas from registry [#16684](https://github.com/apache/druid/pull/16684)
- Fixed issues related to partitioning boundaries in the MSQ task engine's window functions [#16729](https://github.com/apache/druid/pull/16729)

> - Fixed a boost column issue causing quantile sketches to incorrectly estimate the number of output partitions to create [#17141](https://github.com/apache/druid/pull/17141)

Reviewer 5: Nit: This is MSQ window function specific, so we can maybe add that to the message: `Fixed a boost column issue causing quantile sketches to incorrectly estimate the number of output partitions to create for window functions in MSQ task engine`

Also, I see this PR also mentioned in the `Other querying improvements` section - is that expected?

Reviewer 2: Nope, it should not be duplicated. Will remove pending the discussion in the other comment.

- Fixed an issue with `ScanQueryFrameProcessor` cursor build not adjusting intervals [#17168](https://github.com/apache/druid/pull/17168)
- Improved worker cancellation for the MSQ task engine to prevent race conditions [#17046](https://github.com/apache/druid/pull/17046)
- Improved memory management to better support multi-threaded workers [#17057](https://github.com/apache/druid/pull/17057)

> - Reduced memory usage when transferring sketches between the MSQ task engine controller and worker [#16269](https://github.com/apache/druid/pull/16269)

Reviewer 6: This is duplicated from mentioned from line 245. Also, a better way to word it might be "Add new serialization of sketches between MSQ controller and worker to reduce memory usage".

- Improved error handling when retrieving Avro schemas from registry [#16684](https://github.com/apache/druid/pull/16684)

> - Fixed issues related to partitioning boundaries in the MSQ task engine's window functions [#16729](https://github.com/apache/druid/pull/16729)

Reviewer 5: This is duplicated from line 247.

Also, it's a nit but a better message might be: `Fixed issues related to partitioning boundaries for window functions in the MSQ task engine`

- Fixed an issue that caused `maxSubqueryBytes` to fail when segments had missing columns [#16619](https://github.com/apache/druid/pull/16619)
- Fixed an issue with the array type selector that caused the array aggregation over window frame to fail [#16653](https://github.com/apache/druid/pull/16653)
- Fixed support for native window queries without a group by clause [#16753](https://github.com/apache/druid/pull/16753)
- Added a query context parameter to enable trasfers of RAC over wire [17150](https://github.com/apache/druid/pull/17150)
- Window functions now support `maxSubqueryBytes` [#16800](https://github.com/apache/druid/pull/16800)
- Fixed an issue with window functions and partitioning [#17141](https://github.com/apache/druid/pull/17141)
- Updated window functions to disallow multi-value dimensions for partitioning [#17036](https://github.com/apache/druid/pull/17036)
- Fixed an issue with casting objects to vector expressions [#17148](https://github.com/apache/druid/pull/17148)
- Added several fixes and improvements to vectorization fallback [#17098](https://github.com/apache/druid/pull/17098), [#17162](https://github.com/apache/druid/pull/17162)

> - You can now configure encoding method for sketches at query time [#17050](https://github.com/apache/druid/pull/17050)

Reviewer 6: This does not seem to be the correct message for this PR.

Reviewer 2: fixed. thanks

### Window functions are GA

[Window functions](https://druid.apache.org/docs/latest/querying/sql-window-functions) are now generally available in Druid's native engine and in the MSQ task engine.

- You no longer need to use the query context `enableWindowing` to use window functions. [#17087](https://github.com/apache/druid/pull/17087)

### Concurrent append and replace GA

Concurrent append and replace is now GA. The feature safely replaces the existing data in an interval of a datasource while new data is being appended to that interval. One of the most common applications of this feature is appending new data (such as with streaming ingestion) to an interval while compaction of that interval is already in progress. 

### Delta Lake improvements

The community extension for Delta Lake has been improved to support [complex types](#delta-lake-complex-types) and [snapshot versions](#delta-lake-snapshot-versions).

### Iceberg improvements

The community extension for Iceberg has been improved. For more information, see [Iceberg improvements](#iceberg-improvements)

### Projections (experimental)

> Druid 31.0.0 includes experimental support for projections in segments. Like materialized views, projections can improve the performance of queries by optimizing the route the query takes when it executes.

Reviewer 7: ok, i gave this a shot, also included some instruction on how to use the feature since it isn't documented yet
```suggestion
Druid 31.0.0 includes experimental support for new feature called projections. Projections are grouped pre-aggregates of a segment that are automatically used at query time to optimize execution for any queries which 'fit' the shape of the projection by reducing both computation and i/o cost by reducing the number of rows which need to be processed. Projections are contained within segments of a datasource, and do increase the segment size, but are also able to share data such as value dictionaries of dictionary encoded columns with the columns of the base segment.

As an experimental feature, projections are not well documented yet, but can be defined for streaming ingestion and 'classic' batch ingestion as part of the `dataSchema`. For example, using the standard wikipedia example:

```   
    "dataSchema": {
      "granularitySpec": {
        ...
      },
      "dataSource": ...,
      "timestampSpec": {
        ...
      },
      "dimensionsSpec": {
        ...
      },
      "projections": [
        {
          "type": "aggregate",
          "name": "channel_page_hourly_distinct_user_added_deleted",
          "groupingColumns": [
            {
              "type": "long",
              "name": "__gran"
            },
            {
              "type": "string",
              "name": "channel"
            },
            {
              "type": "string",
              "name": "page"
            }
          ],
          "virtualColumns": [
            {
              "type": "expression",
              "expression": "timestamp_floor(__time, 'PT1H')",
              "name": "__gran",
              "outputType": "LONG"
            }
          ],
          "aggregators": [
            {
              "type": "HLLSketchBuild",
              "name": "distinct_users",
              "fieldName": "user",
              "round": true
            },
            {
              "type": "longSum",
              "name": "sum_added",
              "fieldName": "added"
            },
            {
              "type": "longSum",
              "name": "sum_deleted",
              "fieldName": "deleted"
            }
          ]
        },
        ...
      ]
    },
    ...
```

The `groupingColumns` define the order which data is sorted in the projection. Instead of explicitly defining granularity like for the base table, it is defined by defining a virtual column; during ingestion the processing logic finds the ‘finest’ granularity virtual column that is a `timestamp_floor` expression and uses it as the `__time` column for the projection. Projections do not need to have a time column defined, in which case they can still match queries that are not grouping on time.

Projections only can currently be defined by classic ingestion, but they can still be used by queries using MSQ or the new Dart engine. Future development will allow projections to be created as part of MSQ based ingestion as well.

There are a few new query context flags which have been added to aid in experimentation with projections. 
* `useProjection` accepts a specific projection name and instructs the query engine that it must use that projection, and will fail the query if the projection does not match the query
* `forceProjections` accepts `true` or `false` and instructs the query engine that it must use a projection, and will fail the query if it cannot find a matching projection
* `noProjections` accpets `true` or `false` and instructs the query engines to not use any projections

We have a lot of plans to continue to improve this feature in the coming releases, but are excited to get it out there so users can begin experimentation since projections can dramatically improve query performance.
```

Reviewer 7: i'm still working on the writeup for a design proposal for this, another option would be to link to that from this since it should contain some of this information

Reviewer 2: I split this up. Part of it is in the highlight section and the details are in the Querying section. Also instead of including the JSON, I linked to it.

This section contains important information about new and existing features.

### Compaction features

Druid now supports the following features:

- Compaction scheduler with greater flexibility and control over when and what to compact.
- MSQ task engine-based auto-compaction for more performant compaction jobs.

For more information, see [Compaction supervisors](#compaction-supervisors-experimental).

> [#16291](https://github.com/apache/druid/pull/16291)

Reviewer 2: ```suggestion
[#16291](https://github.com/apache/druid/pull/16291) [#16768](https://github.com/apache/druid/pull/16768)
```

The `groupingColumns` field in the spec defines the order which data is sorted in the projection. Instead of explicitly defining granularity like for the base table, you define it with a virtual column. During ingestion, the processing logic finds the ‘finest’ granularity virtual column that is a `timestamp_floor` expression and uses it as the `__time` column for the projection. Projections do not need to have a time column defined. In these cases, they can still match queries that are not grouping on time.

There are new query context flags that have been added to aid in experimentation with projections:

- `useProjection` accepts a specific projection name and instructs the query engine that it must use that projection, and will fail the query if the projection does not match the query
- `forceProjections` accepts true or false and instructs the query engine that it must use a projection, and will fail the query if it cannot find a matching projection

> - `noProjections` accpets true or false and instructs the query engines to not use any projections

Reviewer 3: ```suggestion
- `noProjections` accepts true or false and instructs the query engines to not use any projections
```

- Fixed an issue with tombstone segments incorrectly appearing unavailable in the Web Console [#17025](https://github.com/apache/druid/pull/17025)

### Data management

#### Compaction

##### Compaction supervisors (experimental)

You can now configure automatic compaction to run as a supervisor rather than as a Coordinator duty. Compaction supervisors provide the following benefits over Coordinator duties:

* Can use the supervisor framework to get information about the auto-compaction, such as status or state
* More easily suspend or resume compaction for a datasource
* Can use either the native compaction engine or the MSQ task engine
* More reactive and submits tasks as soon as a compaction slot is available
* Tracked compaction task status to avoid re-compacting an interval repeatedly

You can submit your existing auto-compaction configs as part of the supervisor spec; the two auto-compaction options share the same syntax.

For more information, see [Automatic compaction](https://druid.apache.org/docs/latest/data-management/automatic-compaction).

> [#16291](https://github.com/apache/druid/pull/16291)

Reviewer 2: ```suggestion
[#16291](https://github.com/apache/druid/pull/16291) [16291](https://github.com/apache/druid/pull/16768)
```

> [#16291](https://github.com/apache/druid/pull/16291)

Reviewer 2: ```suggestion
[#16291](https://github.com/apache/druid/pull/16291) [#16768](https://github.com/apache/druid/pull/16768)
```

[#16659](https://github.com/apache/druid/pull/16659)

Hard-coded `"legacy":false` following removal of the legacy mode to prevent error during rolling upgrades or downgrades.

[#16793](https://github.com/apache/druid/pull/16793)

#### ZK-based segment loading

ZK-based segment loading is now disabled. ZK `servedSegmentsPath` was deprecated in Druid 0.7.1. This legacy path has been replaced by `liveSegmentsPath`.

Segment-serving processes such as Peons, Historicals and Indexers no longer create ZK `loadQueuePath` entries. The `druid.zk.paths.loadQueuePath` and `druid.zk.paths.servedSegmentsPath` properties are no longer used.

Move to HTTP-based segment loading first and then perform the version upgrade.

### Developer notes

- Fixed the `hibernate-validator` warning displayed during maven build [#16746](https://github.com/apache/druid/pull/16746)
- Replaced the `StorageAdapter` interface with the `CursorHolder` interface [#17024](https://github.com/apache/druid/pull/17024)

#### Dependency updates

Bumped the versions of the following dependencies:

- Apache Calcite to 1.37.0 [#16621](https://github.com/apache/druid/pull/16621)

> - Delta Kernel from 3.1.0 to 3.2.0 [#16513](https://github.com/apache/druid/pull/16513)

Reviewer 4: This change is superseded by https://github.com/apache/druid/pull/17179, so it should say 3.2.1 instead of 3.2.0

- MySQL Java connector to 8.2.0 [#16024](https://github.com/apache/druid/pull/16024)
- OpenRewrite maven plugin from 5.27.0 to 5.31.0  [#16477](https://github.com/apache/druid/pull/16477)
- Scala library from 2.13.11 to 2.13.14 [#16364](https://github.com/apache/druid/pull/16364)
- Apache Curator reverted to 5.3.0 from 5.5.0 [#16688](https://github.com/apache/druid/pull/16688)
- Updated Blueprint dependencies from v4 to v5 in web console [#16756](https://github.com/apache/druid/pull/16756)
- `io.grpc:grpc-netty-shaded` from 1.57.2 to 1.65.1 [#16731](https://github.com/apache/druid/pull/16731)
- `jclouds.version` from 2.5.0 to 2.6.0 [#16796](https://github.com/apache/druid/pull/16796)
- Axios to 1.7.4 [#16898](https://github.com/apache/druid/pull/16898)
- Delta Kernel to 3.2.1 [#17179](https://github.com/apache/druid/pull/17179)

Other dependency updates:

> - Added support for delta structs, arrays, and maps to the `delta` input source [#16884](https://github.com/apache/druid/pull/16884)

Reviewer 4: This is not really a "dependency", did we mean to say "extension updates"?
